### PR TITLE
[RDY] test: migrate legacy eval test.

### PR DESCRIPTION
**Pull request by lucc**
The original test file did a `:wq` somewhere in the middle.  I "ressurected" the tests that where in test_eval.in after this point but I had to make up the results (they where not in test_eval.ok).  The commit for this is the last one (currently 60b1139).

### CODE REVIEW

**Comment by lucc** [comment]
This is now RFC. The resurrection commit is now 91b9f70.

File: test/functional/legacy/eval_spec.lua [hunk 1]
```diff
@@ -0,0 +1,667 @@
+-- Test for various eval features.
+-- Note: system clipboard support is not tested. I do not think anybody will
+-- thank me for messing with clipboard.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+local eq, eval, wait, write_file = helpers.eq, helpers.eval, helpers.wait, helpers.write_file
+
+describe('various eval features', function()
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: just `'eval'` is enough

File: test/functional/legacy/eval_spec.lua [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+      ]])
+  end)
+  before_each(clear)
+  teardown(function()
+    os.remove('test.out')
+    os.remove('test_eval_func.vim')
+    os.remove('test_eval_setup.vim')
+  end)
+
+  it('let tests', function()
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: just `':let'` is enough.

File: test/functional/legacy/eval_spec.lua [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+    execute('  func! g:test()')
+    execute('    echo "test"')
+    execute('  endfunc')
+    execute('catch')
+    execute('  let tmp = v:exception')
+    execute('endtry')
+    eq('Vim(function):E128: Function name must start with a capital or "s:": g:test()', eval('tmp'))
+  end)
+
+  it('Function name folowed by #', function()
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: typo

**Comment by justinmk** [comment]
@lucc Just curious, did we somehow delete these tests at some point (not sure what you mean by resurrect)?

This LGTM, after squash I will do a fuzz-test locally then merge.

**Comment by lucc** [comment]
I added three more commits. One to apply the comments above. The next two are copying over changes that where made to the original eval test since I started this branch.

About the resurrect question: The original test [exits at line 168](https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_eval.in#L168). The code afterwards is not used and there is no corresponding data in test_eval.ok (the last part in the ok file is about setreg errors).

I used this previously (in vim and nvim) unused test code and "resurrected" it to [lines 599-697 in the new test](https://github.com/lucc/neovim/blob/legacy/eval/test/functional/legacy/eval_spec.lua#L599-L697). I had to think about the expected results myself, please double check them if you have not done so already.

I will now start to squash this but I will put a backup branch with all these WIP commits up on my account so that it can still be discussed if need be.

**Comment by lucc** [comment]
The backup of the WIP commits is [here](https://github.com/lucc/neovim/tree/legacy/eval.presquash.bak).

**Comment by lucc** [comment]
I consider it RDY.

**Comment by bfredl** [comment]
Is there a purpose here to really test the system clipboard integration, or just to test eval stuff that happens to involve `"*` register? If the later, the clipboard can be virtualized using
```
    execute('let &rtp = "test/functional/fixtures,".&rtp')
```
(should be placed right after a `clear()` )

**Comment by lucc** [comment]
@bfredl I did not spend a thought on this. I just tried to migrate *everything* that was in the legacy test. The old eval test tried to test the (real) clipbord integration as I read from the test file and the hg log of the vim repository (`hg log -v --patch src/testdir/test_eval*`).

We could replace the `if has_clipboard()` and `pending()` stuff by just the virtualized clipboard. Actually I think that this test does not test if the yanked text ends up in the system clipboard. It just tests if we can set and get from the `*` register, no check outside of the nvim process is made if the system clipboard was set. I hope that a actual test for a change in the system clipboard is made in `test/functional/clipboard/*`.

File: test/functional/legacy/eval_spec.lua [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+          if a:1 isnot# '='
+              execute "silent normal! Go==\n==\e\"".a:1."P"
+          endif
+      endfun
+      ]])
+  end)
+  before_each(clear)
+  teardown(function()
+    os.remove('test.out')
+    os.remove('test_eval_func.vim')
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: I do not see where this is ever written

File: test/functional/legacy/eval_spec.lua [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+      /: type v; value: abc/]].."\x00 (['abc/\x00']), expr: abc/\x00 (['abc/\x00"..[['])
+      ==
+      =abc/]]..'\x00'..[[=
+      {{{2 setreg('=', ['"abc/"'])
+      =: type v; value: abc/ (['abc/']), expr: "abc/" (['"abc/"'])
+      {{{2 setreg('=', ['"abc/]]..'\x00'..[["'])
+      =: type v; value: abc/]].."\x00 (['abc/\x00"..[[']), expr: "abc/]]..'\x00'..[[" (['"abc/]]..'\x00'..[["'])]])
+  end)
+
+  -- If the clipboard feature is not available the next test will be sciped.
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: sciped => skipped. But the comment is actually not needed.

File: test/functional/legacy/eval_spec.lua [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+    write_file('test_eval_setup.vim', [[
+      set noswapfile
+      lang C
+
+      fun AppendRegContents(reg)
+	  call AppendRegParts(a:reg, getregtype(a:reg), getreg(a:reg), string(getreg(a:reg, 0, 1)), getreg(a:reg, 1), string(getreg(a:reg, 1, 1)))
+      endfun
+
+      fun AppendRegParts(reg, type, cont, strcont, cont1, strcont1)
+	  call append('$', printf('%s: type %s; value: %s (%s), expr: %s (%s)', a:reg, a:type, a:cont, a:strcont, a:cont1, a:strcont1))
```
Comment (justinmk) [inline test/functional/legacy/eval_spec.lua]: convert tab indent to spaces

**Comment by justinmk** [comment]
Thanks @lucc

**Pull request merged**